### PR TITLE
Silence warnings in win32/win32.c

### PR DESCRIPTION
**Pull request by sisyphus**
In win32/win32.c, this PR silences numerous <code>[-Wunused-parameter]</code> and <code>[-Wunused-variable]</code> warnings; also a <code>[-Wsign-compare]</code>, a <code>[-Wdangling-else]</code> and a <code>[-Wparentheses]</code> warning.

The following warnings remain:
```
win32.c: In function 'set_w32_module_name':
win32.c:243:9: warning: cast between incompatible function types from 'FARPROC' {aka 'long long int (*)()'} to 'DWORD (*)(const WCHAR *, WCHAR *, DWORD)' {aka 'long unsigned int (*)(const short unsigned int *, short unsigned int *, long unsigned int)'} [-Wcast-function-type]
  243 |         (DWORD (__stdcall *)(LPCWSTR, LPWSTR, DWORD))
      |         ^
win32.c: In function 'win32_stat_low':
win32.c:1688:25: warning: cast between incompatible function types from 'FARPROC' {aka 'long long int (*)()'} to 'DWORD (*)(void *, CHAR *, DWORD,  DWORD)' {aka 'long unsigned int (*)(void *, char *, long unsigned int,  long unsigned int)'} [-Wcast-function-type]
 1688 |                         (pGetFinalPathNameByHandleA_t)GetProcAddress(GetModuleHandle("kernel32.dll"), "GetFinalPathNameByHandleA");
      |                         ^
win32.c: In function 'win32_link':
win32.c:3708:40: warning: passing argument 1 to 'restrict'-qualified parameter aliases with argument 2 [-Wrestrict]
 3708 |         ((aTHXa(PERL_GET_THX)), wcscpy(wOldName, PerlDir_mapW(wOldName)),
      |                                        ^~~~~~~~
win32.c: In function 'win32_symlink':
win32.c:3733:9: warning: cast between incompatible function types from 'FARPROC' {aka 'long long int (*)()'} to 'BOOLEAN (*)(const CHAR *, const CHAR *, DWORD)' {aka 'unsigned char (*)(const char *, const char *, long unsigned int)'} [-Wcast-function-type]
 3733 |         (pCreateSymbolicLinkA_t)GetProcAddress(GetModuleHandle("kernel32.dll"), "CreateSymbolicLinkA");
      |         ^
win32.c: In function 'win32_hook_closehandle_in_crt':
win32.c:5370:24: warning: cast between incompatible function types from 'FARPROC' {aka 'long long int (*)()'} to 'BOOL (*)(void *)' {aka 'int (*)(void *)'} [-Wcast-function-type]
 5370 |     CloseHandle_orig = (pCloseHandle)GetProcAddress(
      |                        ^
win32.c:5377:36: warning: cast between incompatible function types from 'BOOL (*)(void *)' {aka 'int (*)(void *)'} to 'INT_PTR (*)()' {aka 'long long int (*)()'} [-Wcast-function-type]
 5377 |         crt_handle, "CloseHandle", (FARPROC)my_CloseHandle
      |                                    ^
win32.c:5380:31: warning: cast between incompatible function types from 'FARPROC' {aka 'long long int (*)()'} to 'NTSTATUS (*)(void *, struct _IO_STATUS_BLOCK *, void *, ULONG,  ULONG)' {aka 'long int (*)(void *, struct _IO_STATUS_BLOCK *, void *, long unsigned int,  long unsigned int)'} [-Wcast-function-type]
 5380 |     pNtQueryInformationFile = (pNtQueryInformationFile_t)GetProcAddress(
      |                               ^
win32.c: In function 'win32_unhook_closehandle_in_crt':
win32.c:5404:36: warning: cast between incompatible function types from 'pCloseHandle' {aka 'int (*)(void *)'} to 'INT_PTR (*)()' {aka 'long long int (*)()'} [-Wcast-function-type]
 5404 |         crt_handle, "CloseHandle", (FARPROC)CloseHandle_orig
      |                                    ^
win32.c: At top level:
win32.c:2250:1: warning: 'wstr_to_str' defined but not used [-Wunused-function]
 2250 | wstr_to_str(const wchar_t* wstr)
      | ^~~~~~~~~~~
....

win32.c: In function 'win32_gets':
win32.c:4742:12: warning: call to 'gets' declared with attribute warning: Using gets() is always unsafe - use fgets() instead [-Wattribute-warning]
 4742 |     return gets(s);
      |            ^~~~~~~
```

I have queries about the best way to deal with those remaining warnings, and I'm thinking if we can just pick the low hanging fruit for now, it will make it easier to deal with those remaining issues later on.

My <code>perl -V</code> output is:
```
Summary of my perl5 (revision 5 version 39 subversion 8) configuration:
  Derived from:
  Platform:
    osname=MSWin32
    osvers=10.0.22621.3007
    archname=MSWin32-x64-multi-thread
    uname=''
    config_args='undef'
    hint=recommended
    useposix=true
    d_sigaction=undef
    useithreads=define
    usemultiplicity=define
    use64bitint=define
    use64bitall=undef
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
  Compiler:
    cc='gcc'
    ccflags ='-Wall -Wextra -Werror=pointer-arith -Wno-format -Wno-long-long -Werror=vla -std=c99 -DWIN32 -DWIN64 -fdiagnostics-color=never -DPERL_TEXTMODE_SCRIPTS -DMULTIPLICITY -DPERL_IMPLICIT_SYS -DUSE_PERLIO -D__USE_MINGW_ANSI_STDIO -fwrapv -fno-strict-aliasing -mms-bitfields'
    optimize='-Os'
    cppflags='-DWIN32'
    ccversion=''
    gccversion='13.2.0'
    gccosandvers=''
    intsize=4
    longsize=4
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='long long'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='g++'
    ldflags ='-s -L"D:\perl-5.39.71-1320-ucrt-mcf\lib\MSWin32-x64-multi-thread\CORE" -L"C:\winlibs-ucrt-1320\mingw64\lib" -L"C:\winlibs-ucrt-1320\mingw64\x86_64-w64-mingw32\lib" -L"C:\winlibs-ucrt-1320\mingw64\lib\gcc\x86_64-w64-mingw32\13.2.0"'
    libpth=C:\winlibs-ucrt-1320\mingw64\lib C:\winlibs-ucrt-1320\mingw64\x86_64-w64-mingw32\lib C:\winlibs-ucrt-1320\mingw64\lib\gcc\x86_64-w64-mingw32\13.2.0 D:\_64\msys_1320\1.0\local\lib
    libs= -lmoldname -lkernel32 -luser32 -lgdi32 -lwinspool -lcomdlg32 -ladvapi32 -lshell32 -lole32 -loleaut32 -lnetapi32 -luuid -lws2_32 -lmpr -lwinmm -lversion -lodbc32 -lodbccp32 -lcomctl32
    perllibs= -lmoldname -lkernel32 -luser32 -lgdi32 -lwinspool -lcomdlg32 -ladvapi32 -lshell32 -lole32 -loleaut32 -lnetapi32 -luuid -lws2_32 -lmpr -lwinmm -lversion -lodbc32 -lodbccp32 -lcomctl32
    libc=-lucrt
    so=dll
    useshrplib=true
    libperl=libperl539.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_win32.xs
    dlext=dll
    d_dlsymun=undef
    ccdlflags=' '
    cccdlflags=' '
    lddlflags='-shared -s -L"D:\perl-5.39.71-1320-ucrt-mcf\lib\MSWin32-x64-multi-thread\CORE" -L"C:\winlibs-ucrt-1320\mingw64\lib" -L"C:\winlibs-ucrt-1320\mingw64\x86_64-w64-mingw32\lib" -L"C:\winlibs-ucrt-1320\mingw64\lib\gcc\x86_64-w64-mingw32\13.2.0"'


Characteristics of this binary (from libperl):
  Compile-time options:
    HAS_LONG_DOUBLE
    HAS_TIMES
    HAVE_INTERP_INTERN
    MULTIPLICITY
    PERLIO_LAYERS
    PERL_COPY_ON_WRITE
    PERL_DONT_CREATE_GVSV
    PERL_HASH_FUNC_SIPHASH13
    PERL_HASH_USE_SBOX32
    PERL_IMPLICIT_SYS
    PERL_MALLOC_WRAP
    PERL_OP_PARENT
    PERL_PRESERVE_IVUV
    PERL_USE_SAFE_PUTENV
    USE_64_BIT_INT
    USE_ITHREADS
    USE_LARGE_FILES
    USE_LOCALE
    USE_LOCALE_COLLATE
    USE_LOCALE_CTYPE
    USE_LOCALE_NUMERIC
    USE_LOCALE_TIME
    USE_PERLIO
    USE_PERL_ATOF
    USE_THREAD_SAFE_LOCALE
  Locally applied patches:
    uncommitted-changes
  Built under MSWin32
  Compiled at Feb  4 2024 22:51:26
  @INC:
    D:/perl-5.39.71-1320-ucrt-mcf/site/lib/MSWin32-x64-multi-thread
    D:/perl-5.39.71-1320-ucrt-mcf/site/lib
    D:/perl-5.39.71-1320-ucrt-mcf/lib/MSWin32-x64-multi-thread
    D:/perl-5.39.71-1320-ucrt-mcf/lib
```

### CODE REVIEW

File: win32/win32.c [hunk 1]
```diff
@@ -1455,7 +1458,7 @@ win32_kill(int pid, int sig)
                     /* kill -9 style un-graceful exit */
                     /* Do a wait to make sure child starts and isn't in DLL
                      * Loader Lock */
-                    HWND hwnd = get_hwnd_delay(aTHX, child, 5);
```
Comment (tonycoz) [inline win32/win32.c:1458]: get_hwnd_delay() processes messages, which we do want here.

This should probably be:
```
(void)get_hwnd_delay(aTHX, child, 5);
```

File: win32/win32.c [hunk 2]
```diff
@@ -5540,6 +5538,8 @@ win32_signal(int sig, Sighandler_t subcode)
 LRESULT CALLBACK
 win32_message_window_proc(HWND hwnd, UINT msg, WPARAM wParam, LPARAM lParam)
 {
+    PERL_UNUSED_ARG(hwnd);
+    PERL_UNUSED_ARG(lParam);
```
Comment (tonycoz) [inline win32/win32.c]: These arguments are used and I don't see a warning for them.

File: win32/win32.c [hunk 3]
```diff
@@ -5013,6 +5008,7 @@ XS(w32_GetCwd)
      *   then it worked, set PV valid,
      *   else return 'undef'
      */
+    PERL_UNUSED_VAR(items);
```
Comment (tonycoz) [inline win32/win32.c]: This is added between a comment and the code it comments.  Better would be immediately after `dXSARGS`

**Comment by tonycoz** [comment]
> ```
> win32.c: In function 'set_w32_module_name':
> win32.c:243:9: warning: cast between incompatible function types from 'FARPROC' {aka 'long long int (*)()'} to 'DWORD (*)(const WCHAR *, WCHAR *, DWORD)' {aka 'long unsigned int (*)(const short unsigned int *, short unsigned int *, long unsigned int)'} [-Wcast-function-type]
>   243 |         (DWORD (__stdcall *)(LPCWSTR, LPWSTR, DWORD))
>       |         ^
> ```

We probably need to add `-Wno-cast-function-type`.

> ```
> win32.c: At top level:
> win32.c:2250:1: warning: 'wstr_to_str' defined but not used [-Wunused-function]
>  2250 | wstr_to_str(const wchar_t* wstr)
>       | ^~~~~~~~~~~
> ```

The code that uses this is conditional on `_DEBUG`, which this could be too.

I'm not sure about gets().

**Comment by sisyphus** [comment]
I've added <code>-Wno-cast-function-type</code> to the flags (in my local build of perl), and made the alterations suggested by @tonycoz.
That reduces the remaining warnings to just 3:
```
win32.c: In function 'win32_link':
win32.c:3708:40: warning: passing argument 1 to 'restrict'-qualified parameter aliases with argument 2 [-Wrestrict]
 3708 |         ((aTHXa(PERL_GET_THX)), wcscpy(wOldName, PerlDir_mapW(wOldName)),

win32.c: At top level:
win32.c:2250:1: warning: 'wstr_to_str' defined but not used [-Wunused-function]
 2250 | wstr_to_str(const wchar_t* wstr)
      | ^~~~~~~~~~~

win32.c: In function 'win32_gets':
win32.c:4742:12: warning: call to 'gets' declared with attribute warning: Using gets() is always unsafe - use fgets() instead [-Wattribute-warning]
 4742 |     return gets(s);
      |            ^~~~~~~
```
Is there anything I should try at this stage to deal with any of those remaining warnings ?
We haven't touched upon the <code>[-Wrestrict]</code> warning, which is included in <code>-Wall</code> in gcc-13.2.0.

**Comment by tonycoz** [comment]
The `-Wrestrict` warning may be a real issue, violating the rules on restrict can result in undefined behaviour.

The definition of wstr_to_str() and probably the declaration at the top should be in `#ifdef _DEBUG`.

The use of gets() could be surrounded by GCC_DIAG_IGNORE_STMT():
```
DllExport char*
win32_gets(char *s)
{
    GCC_DIAG_IGNORE_STMT(-Wattribute-warning); /* gets() only for compat */
    return gets(s);
    GCC_DIAG_RESTORE_STMT;
}
```

**Comment by sisyphus** [comment]
Having built perl with <code>-Wno-cast-function-type</code> added to the compilation flags, I now see only the <code>[-Wrestrict]</code> warning issued by win32_link().
The mingw-w64 CI build, which doesn't have that extra flag, still emits the <code>[-Wcast-function-type]</code> warnings.

**Comment by sisyphus** [comment]
I have prepared other files to avoid issuing of pointless warnings - including the removal of 650 lines of such warnings from win32/perlhost.h.
I was thinking it might be better to first get this current PR merged.
But I could add these other files to this PR if that's deemed preferable.
How should I proceed ?

**Comment by tonycoz** [comment]
Please squash the 3 commits here into one commit.

Please add the perlhost fixes to a new PR.

Thanks

**Comment by sisyphus** [comment]
I think I've succeeded in squashing the 3 commits into 1.
Anything else needed ?

**Review by tonycoz — APPROVED** [review]

**Pull request merged**